File: Semester5/Parallel and Distributed Programming/Tema1/Java/LabJava/src/Columns/ColumnsMethod.java
package Columns;

import Rows.RowsThread;

public class ColumnsMethod {
    private final int[][] matrix;
    private final int[][] kernel;
    private final int N;
    private final int M;
    private final int p;

    public ColumnsMethod(int[][] matrix, int[][] kernel, int p) {
        this.matrix = matrix;
        this.kernel = kernel;
        N = matrix.length;
        M = matrix[0].length;
        this.p = p;
    }

    public int[][] start() {
        Thread[] threads = new ColumnsThread[p];

        int start = 0;
        int end = M / p;
        int rest = M % p;

        int[][] result = new int[N][M];

        for (int i = 0; i < p; i++) {
            if (rest > 0) {
                end++;
                rest--;
            }
            threads[i] = new ColumnsThread(matrix, kernel, result, start, end);
            threads[i].start();

            start = end;
            end = start + M / p;
        }

        for (int i = 0; i < p; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }

        return result;
    }
}

File: Semester5/Parallel and Distributed Programming/Tema1/C++/CppDinamic/CppDinamic/ColumnsMethod.cpp
#include "ColumnsMethod.h"
#include "Utils.h"
#include <vector>
#include <thread>

void ColumnsMethod::task(std::vector<std::vector<int>>& result, int start, int end)
{
    int N = matrix.size();  
    for (int i = 0; i < N; i++) {
        for (int j = start; j < end; j++) {
            result[i][j] = Utils::convolveAt(matrix, kernel, i, j); 
        }
    }
}


ColumnsMethod::ColumnsMethod(const std::vector<std::vector<int>>& matrix, const std::vector<std::vector<int>>& kernel, int p)
    : matrix(matrix), kernel(kernel), p(p) {}

void ColumnsMethod::start(std::vector<std::vector<int>>& result)
{
    int N = matrix.size();  
    int M = matrix[0].size();  
    std::vector<std::thread> threads(p); 

    int start = 0;
    int end = M / p;
    int rest = M % p;

    for (int i = 0; i < p; i++) {
        if (rest > 0) {
            end++;  
            rest--;
        }

        threads[i] = std::thread(&ColumnsMethod::task, this, std::ref(result), start, end);

        start = end;
        end = start + M / p;
    }

    for (int i = 0; i < p; i++) {
        threads[i].join();
    }
}
